source/material_model/steinberger.cc

```diff
@@ -392,166 +219,54 @@ namespace aspect
-      fill_mass_and_volume_fractions (in, mass_fractions, volume_fractions);
+      equation_of_state.fill_mass_and_volume_fractions (in, mass_fractions, volume_fractions);
 
       for (unsigned int i=0; i < in.n_evaluation_points(); ++i)
         {
           if (in.requests_property(MaterialProperties::viscosity))
             out.viscosities[i] = viscosity(in.temperature[i], in.pressure[i], in.composition[i], in.strain_rate[i], in.position[i]);
 
           out.thermal_conductivities[i] = thermal_conductivity_value;
-          out.entropy_derivative_pressure[i]    = 0;
-          out.entropy_derivative_temperature[i] = 0;
           for (unsigned int c=0; c<in.composition[i].size(); ++c)
             out.reaction_terms[i][c]            = 0;
 
-          // The following lines take the appropriate averages of the
-          // thermodynamic material properties
-          std::vector<double> specific_heats(material_lookup.size());
-          std::vector<double> densities(material_lookup.size());
-          std::vector<double> compressibilities(material_lookup.size());
-          std::vector<double> thermal_expansivities(material_lookup.size());
-
-          for (unsigned int j=0; j<material_lookup.size(); ++j)
-            {
-              densities[j] = material_lookup[j]->density(in.temperature[i],in.pressure[i]);
-              compressibilities[j] = material_lookup[j]->dRhodp(in.temperature[i],in.pressure[i])/densities[j];
-
-              if (!latent_heat)
-                {
-                  thermal_expansivities[j] = material_lookup[j]->thermal_expansivity(in.temperature[i],in.pressure[i]);
-                  specific_heats[j] = material_lookup[j]->specific_heat(in.temperature[i],in.pressure[i]);
-                }
-            }
+          // Evaluate the equation of state properties at the current evaluation point
+          equation_of_state.evaluate(in, i, eos_outputs);
 
           // The density and isothermal compressibility are both volume-averaged
-          out.densities[i] = MaterialUtilities::average_value(volume_fractions[i], densities, MaterialUtilities::arithmetic);
-          out.compressibilities[i] = MaterialUtilities::average_value(volume_fractions[i], compressibilities, MaterialUtilities::arithmetic);
+          out.densities[i] = MaterialUtilities::average_value (volume_fractions[i], eos_outputs.densities, MaterialUtilities::arithmetic);
+          out.compressibilities[i] = MaterialUtilities::average_value (volume_fractions[i], eos_outputs.compressibilities, MaterialUtilities::arithmetic);
+          out.entropy_derivative_pressure[i] = MaterialUtilities::average_value (mass_fractions[i], eos_outputs.entropy_derivative_pressure, MaterialUtilities::arithmetic);
+          out.entropy_derivative_temperature[i] = MaterialUtilities::average_value (mass_fractions[i], eos_outputs.entropy_derivative_temperature, MaterialUtilities::arithmetic);
 
           if (!latent_heat)
```

Comment:
The if statement was to avoid unnecessary calculations. I've now migrated all of this code block to the thermodynamic lookup, so it won't be a problem any more.